### ISSUE TITLE
UTF8 & WL issues

### ISSUE THREAD

**Issue by kkkmail** [halirutan/Mathematica-IntelliJ-Plugin#92]
[UTF8_issue.zip](https://github.com/halirutan/Mathematica-IntelliJ-Plugin/files/1404522/UTF8_issue.zip)

Plug in cannot properly handle UTF8 M files and cannot handle WL files with UTF8 characters. 

In fact, it looks that plug in cannot handle standard WL files at all! If an NB file is saved as WL, then it comes out with each line effectively commented out. While Mathematica seems to handle that correctly, plug in shows it as a file with each line commented out! 

If NB file is saved as a text file (but with M extension), then some of the characters might get UTF8 encoding and some might just get into HEX representation, like in UTF8_issue_ASCII.m. 

**To sum it up**: Mathematica can handle correctly all 3 different versions of the file: **UTF8_issue.m**, **UTF8_issue.wl**, **UTF8_issue_ASCII.m** (the last one is a pure ASCII version). Plug in cannot handle any of them properly.

**Comment by halirutan** [comment]
No, it can not and I won't support this. Mathematica already has a tremendous amount of operators in plain ASCII that need to the parsed and transformed into an AST to make the plugin do what it does. Supporting all the nifty symbols will make this bottleneck even worse and it will be used by 0.1% of package developers. Additionally, it will make things harder for people that have to read your code.
Let me give some examples that show how bad this is and why I will always be against this.

Your arrow from the examples looks very similar to the `Rule` arrow. This is by far not the only instance but it can lead to errors that are hard to find

![Mathematica graphics](http://i.stack.imgur.com/V261b.png)

The `TensorProduct` UTF8 char does not render correctly in no editor or terminal I have. Not even in the latest Workbench. Even worse, it is not recognized as the operator it is. So this in WB

![wb](http://i.imgur.com/V32d7Ih.png)

is not recognized as tensor product but it is a symbol name that starts with `matrix` and ends with `matrix`. You have to understand that if an UTF8 character is a valid operator, it needs to be handled like one and we need its exact precedence in comparison to other operators. Otherwise, the whole plugin goes down the drain.

Another example that looks innocent but two different plus signs are used:

![ex](http://i.imgur.com/FaQsweM.png)

Even if this is the correct UTF8 for [the plus sign](http://www.fileformat.info/info/unicode/char/2b/index.htm), Mathematica handles it as it were a character for a symbol

![Mathematica graphics](http://i.stack.imgur.com/1U3ks.png)

The amount of work to support something like this correctly is just not justified by the bad outcome. But since the plugin is open-source, feel free to fork it and include this. My opinion is that package code should be plain ASCII.

**Comment by kkkmail** [comment]
That's exactly my point: plug in cannot handle plain ASCII! See UTF8_issue_ASCII.m and try to open it in IntelliJ. The only thing that plug in needs to "know" is that things like `\:0df4` are characters.

**Comment by halirutan** [comment]
Yes, you already reported this here (https://github.com/halirutan/Mathematica-IntelliJ-Plugin/issues/65) and I tried to explain that it doesn't make a difference from the plugin's point of view. Still, it would need to parse every operator in possibly 4 different forms. Here is a specific example

```
varA∶varB
varA\[Colon]varB
varA\:2236varB
Colon[varA, varB]
```

Evaluate this code in Mathematica and you see that all 4 forms are equal. The precedence of `Colon` is 80, the precedence of Plus is 310. So Plus binds stronger

```
varA∶varB+varC
varA\[Colon]varB+varC
varA\:2236varB+varC
```

Use this in Mathematica and double click on `varB`. You'll notice it expands to the right because it binds stronger. Workbench does this completely wrong and I don't blame it for that. Here

```
varA∶varB+varC
```

`varA∶varB` is recognized as one symbol. Here

```
varA\[Colon]varB+varC
varA\:2236varB+varC
```
`Colon` is correctly identified as operator, but binds stronger than Plus. The main goal of the plugin is to help to code. For this, I need to identify and find local variables and global definitions. Especially for the last part, I must ensure that I have the correct syntax tree of the code. Supporting something like that makes it by magnitudes harder. Consider this

![Mathematica graphics](http://i.stack.imgur.com/bP3ru.png)

By not supporting those characters, you are forced to write this in IDEA

```
varA /: Colon[varA, varB[x_]] := x
```
That makes it crystal clear to the engine how string `Colon` binds because you had to put brackets and in Mathematica it will happily render to

![Mathematica graphics](http://i.stack.imgur.com/iEvoe.png)

**Comment by kkkmail** [comment]
Thanks. I see your point. I guess that I fall into this 0.1% of users and I am stuck. I explicitly looked through tens of thousands of Unicode characters available in Mathematica to come up with easily human readable things like:
`coeff෴A˖A➔Y˖A`, which are parsed as variable names. It describes a rate of a chemical reaction (that's why `coeff `prefix) and `A` defines an amino acid in a peptide chain (up to 47 different ones). So if could be something as bad as: `coeff෴B˖AαCΘ➔Y˖AαCΘ` (the upper case and lower case denote so called enantiomers of the amino acids and English and Greek letters are the only ones, which "understand" upper and lower case). I am not too good in java, so I can't really contribute enough to the project. But anyway, thanks for your time and efforts.

**Comment by kubaPod** [comment]
@kkkmail 

I just want to comment on 

> In fact, it looks that plug in cannot handle standard WL files at all! If an NB file is saved as WL, then it comes out with each line effectively commented out. While Mathematica seems to handle that correctly, plug in shows it as a file with each line commented out!

which is not related to UTF8 support but I want to make that clear, sorry if I missed the point.

Only cells with `InitializationCell -> True` option won't be commented when a notebook is converted to a .m file. `"Code"` cells by default and all those for which you set this option manually/via menu, everything else, like Input etc is commented. 

During import MMA correctly parses those comments back but expecting the plugin to do this would be analogous to except from markdown plugins to allow edition in the view mode. Ok, not the best analogy but the point is, it is not plugin's job.

See more in http://reference.wolfram.com/language/ref/AutoGeneratedPackage.html
